Scetch a possiblity for custom logic in the broker

flxo opened the pull request:
Use the authentication stuff as an example how customer code could be plugged into the broker. The authentification is not implemented - it's all just about how to proceed...

The concrete type that impls the authentication has the advantage of static dispatch. There ware probably other traits needed for intercepting publishes and subscriptions e.g 

```rust
pub trait {
  async fn on_publish(client_id: &str, publish: &PublishPacket) -> PublicationResult;
  ....
}


Do *not* merge this :-)

flxo committed:
Scetch a possiblity for custom logic in the broker

Some thoughts about a broker plugins

flxo commented:
@bschwind a quick attempt how to plug custom logic into the broker - no logic implemented - just to get the discussion started. Thanks for comments!

flxo committed:
Add subscribe and publish to plugin trait

flxo committed:
Attempt to implement authentification

bschwind commented:
Hey @flxo, sorry for the late response! Been on holiday the past week, but I'll try to take a look at this soon.

flxo commented:
@bschwind Thanks!

bschwind commented on Cargo.lock:
```diff
@@ -24,15 +24,18 @@ dependencies = [
 
```
Was this Cargo.lock update needed?

flxo replied:
No. Fixed.

bschwind commented on mqtt-v5-broker/src/plugin.rs:
```diff
@@ -0,0 +1,124 @@
+    fn on_subscribe(&mut self, packet: &SubscribePacket) -> SubscribeAckPacket;
+
+    /// Called on publish packets reception for QoS 0. Return true if the packet should be published to the clients.
+    fn on_publish_received_qos0(&mut self, packet: &PublishPacket) -> bool;
```
What's the reasoning for splitting out publishes based on the qos? `PublishPacket` already has the `qos` field on it, but maybe there's some other reasoning for it that I'm missing.

Is this for handling the individual steps in the QoS sending handshake, or just a wholesale "We received a qos2 packet, do you want to send it?"

flxo replied:
I had a single "callback" for publication but struggled with the return type. There's no return for QOS0, a Publish Ack for QOS1 and PublishReceivedAck for QOS2. 
I'm fine with merging those if you want and but would require to define a struct for that the includes basically the same as the `PublishAckPacket` and `PublishReceivedPacket` in an `Option`.

bschwind commented on mqtt-v5-broker/src/broker.rs:
```diff
@@ -45,14 +55,33 @@ struct Session {
     will: Option<FinalWill>,
 }
 
+impl std::fmt::Debug for Session {
```
Was `pub client_sender: Option<Sender<ClientMessage>>,` giving you trouble?

I'd prefer to avoid a manual `Debug` impl which is only working around one particular field, probably better to fix the field itself with a new-type, or see if there's an alternative which _does_ impl `Debug`.

flxo replied:
There's no trouble - it's just noisy  and the internals of the `Sender` are not interesting in the context of MQTT. It's a `trace` so I switch it to a `Debug` derive.

bschwind commented on mqtt-v5-broker/src/broker.rs:
```diff
@@ -190,24 +223,58 @@ pub enum BrokerMessage {
 }
 
-pub struct Broker {
+pub struct Broker<A> {
```
Generic arguments can have a default value so maybe specifying our default plugin impl would be good here: 

`pub struct Broker<A = Noop>`

bschwind commented on mqtt-v5-broker/src/broker.rs:
```diff
@@ -178,7 +209,9 @@ pub enum WillDisconnectLogic {
 
 #[derive(Debug)]
 pub enum BrokerMessage {
+    Stats,
```
This feels a bit unrelated to the topic of introducing plugins to the broker, could you separate it from this PR?

flxo replied:
Sure. I removed it.

For metrics I would sugest  [prometheus](https://prometheus.io) - but that's as you noted unrelated to this discussion.

bschwind reviewed (commented):
@flxo thanks for the work so far! I saw you originally were using async trait methods (via the `async-trait` crate), what was your reasoning for switching away from it?

It's not that I'm for or against it, just curious about the reasoning. I think that will be a big decision for us - do we want to allow async operations in the plugins? I can imagine a plugin that would want to read from a database or key-value store before making a decision, so it _might_ be nicer to have the functions be async. I haven't thought through all the implications yet.

I looked less at the logic changes in `broker.rs` as I think settling on the API design first is important.

I left some more targeted comments in the code itself. Thanks again!

flxo commented:
> @flxo thanks for the work so far! I saw you originally were using async trait methods (via the `async-trait` crate), what was your reasoning for switching away from it?
> 
> It's not that I'm for or against it, just curious about the reasoning. I think that will be a big decision for us - do we want to allow async operations in the plugins? I can imagine a plugin that would want to read from a database or key-value store before making a decision, so it _might_ be nicer to have the functions be async. I haven't thought through all the implications yet.
> 
> I looked less at the logic changes in `broker.rs` as I think settling on the API design first is important.
> 
> I left some more targeted comments in the code itself. Thanks again!

The reason why I used the `async-trait` thing is because I need exactly what you described: query an eternal entity where I have a client for that also uses tokio and provides async fn. I haven't look into cool ways to call async code from sync without channels in between.

I will try to answer your findings inline.

flxo committed:
Remove debug stats

flxo committed:
Add client disconnect callback

bschwind commented:
@flxo still keeping an eye on this PR! Sorry though that I don't have the bandwidth for quick reviews. I'll be happy to give it a look when you think it's ready for another review.

flxo commented:
> @flxo still keeping an eye on this PR! Sorry though that I don't have the bandwidth for quick reviews. I'll be happy to give it a look when you think it's ready for another review.

Thanks. I'll try to solve some of the todos I left in the PR and will ping you.

flxo committed:
Refactor authentification handling

Collect unauthorized connections seperated from the session until the
auth is complete. Introduce a unique identifier for connections.

flxo committed:
Optimize connection id

Replace String connection id with cheaper u64. The connection id is a
atomically inremented counter in the client handler.

flxo committed:
Rename UnauthenticatedSession to UnauthenticatedConnection

flxo commented:
@bschwind I refactored the connection handling a bit and separated the unauthenticated connections from the session. Introduced a `ConnectionId` which is a unique identifier for a connection.

I'd love to hear your feedback.

bschwind commented:
Hi @flxo 

Sorry for the long time radio silence. I still think this is good work, but I don't have enough bandwidth to review this in any sort of timely fashion.

I plan to get back into this project around the time when I update my ESP32 projects to use Rust, but until then you probably won't see too much activity from me on it. I would recommend working with either your fork of this repo, or rumqtt (which I would guess is farther along and has more contributors)
